### ISSUE TITLE
test_names.ServerDNSTestCase.tearDown doesn't wait for TCP Port closure

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2149]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| PenguinOfDoom reported|
|-|-|
|Trac ID|trac#2149|
|Type|defect|
|Created|2006-10-03 23:05:07Z|

Here's a patch:

```
Index: twisted/names/test/test_names.py
===================================================================
--- twisted/names/test/test_names.py    (revision 18379)
+++ twisted/names/test/test_names.py    (working copy)
@@ -152,8 +152,9 @@
         self.resolver = client.Resolver(servers=[('127.0.0.1', port)])

     def tearDown(self):
-        self.listenerTCP.loseConnection()
-        d = defer.maybeDeferred(self.listenerUDP.stopListening)
+        d1 = self.listenerTCP.loseConnection()
+        d2 = defer.maybeDeferred(self.listenerUDP.stopListening)
+        d = defer.gatherResults([d1, d2])
         def disconnectTransport(ignored):
             if getattr(self.resolver.protocol, 'transport', None) is not None:
                 return self.resolver.protocol.transport.stopListening()
```

<details><summary>Searchable metadata</summary>

```
trac-id__2149 2149
type__defect defect
reporter__PenguinOfDoom PenguinOfDoom
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1159916707000000 1159916707000000
changetime__1160245249000000 1160245249000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_1"></a>|@jerub set owner to PenguinOfDoom|
|-|-|

Please add a docstring to this method. Once you do that, reassign it to me for review.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|PenguinOfDoom set owner to @jerub|
|-|-|

How about this:

```
Index: test_names.py
===================================================================
--- test_names.py       (revision 18379)
+++ test_names.py       (working copy)
@@ -152,8 +152,10 @@
         self.resolver = client.Resolver(servers=[('127.0.0.1', port)])
 
     def tearDown(self):
-        self.listenerTCP.loseConnection()
-        d = defer.maybeDeferred(self.listenerUDP.stopListening)
+        """Asynchronously disconnect listenerTCP, listenerUDP and the resolver"""
+        d1 = self.listenerTCP.loseConnection()
+        d2 = defer.maybeDeferred(self.listenerUDP.stopListening)
+        d = defer.gatherResults([d1, d2])
         def disconnectTransport(ignored):
             if getattr(self.resolver.protocol, 'transport', None) is not None:
                 return self.resolver.protocol.transport.stopListening()
```

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|PenguinOfDoom commented|
|-|-|

Here's a docstring nazi special:

```
Index: test_names.py
===================================================================
--- test_names.py       (revision 18379)
+++ test_names.py       (working copy)
@@ -152,8 +152,10 @@
         self.resolver = client.Resolver(servers=[('127.0.0.1', port)])
 
     def tearDown(self):
-        self.listenerTCP.loseConnection()
-        d = defer.maybeDeferred(self.listenerUDP.stopListening)
+        """Asynchronously disconnect listenerTCP, listenerUDP and resolver"""
+        d1 = self.listenerTCP.loseConnection()
+        d2 = defer.maybeDeferred(self.listenerUDP.stopListening)
+        d = defer.gatherResults([d1, d2])
         def disconnectTransport(ignored):
             if getattr(self.resolver.protocol, 'transport', None) is not None:
                 return self.resolver.protocol.transport.stopListening()
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_5"></a>|@jerub set owner to PenguinOfDoom|
|-|-|

please merge

**Comment by twisted-trac** [comment]
|<img alt="pahan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|pahan set status to `closed`|
|-|-|

(In [18402]) Improve cleanup for ServerDNSTestCase

Author: pahan
Reviewer: jerub
Fixes [#2149](https://github.com/twisted/twisted/issues/2149)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
